syft/pkg/catalog.go

```diff
@@ -41,43 +39,133 @@ func (c *Catalog) PackageCount() int {
 
 // Package returns the package with the given ID.
 func (c *Catalog) Package(id ID) *Package {
-	return c.byID[id]
+	v, exists := c.byID[id]
+	if !exists {
+		return nil
+	}
+	return v
+}
+
+// PackagesByPath returns all packages that were discovered from the given path.
+func (c *Catalog) PackagesByPath(path string) []*Package {
+	return c.Packages(c.idsByPath[path])
 }
 
-// PackagesByFile returns all packages that were discovered from the given source file reference.
-func (c *Catalog) PackagesByFile(location source.Location) []*Package {
-	return c.byFile[location]
+// Packages returns all packages for the given ID.
+func (c *Catalog) Packages(ids []ID) (result []*Package) {
+	for _, i := range ids {
+		p, exists := c.byID[i]
+		if exists {
+			result = append(result, p)
+		}
+	}
+	return result
 }
 
 // Add a package to the Catalog.
 func (c *Catalog) Add(p Package) {
-	if p.id != 0 {
-		log.Errorf("package already added to catalog: %s", p)
-		return
-	}
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	p.id = ID(nextPackageID)
-	nextPackageID++
+	_, exists := c.byID[p.ID]
+	if exists {
+		log.Errorf("package ID already exists in the catalog : id=%+v %+v", p.ID, p)
+		return
+	}
+
+	if p.ID == "" {
+		p.ID = newID()
+	}
 
 	// store by package ID
-	c.byID[p.id] = &p
+	c.byID[p.ID] = &p
 
 	// store by package type
-	_, ok := c.byType[p.Type]
-	if !ok {
-		c.byType[p.Type] = make([]*Package, 0)
+	c.idsByType[p.Type] = append(c.idsByType[p.Type], p.ID)
+
+	// store by file location paths
+	for _, l := range p.Locations {
+		if l.RealPath != "" {
+			c.idsByPath[l.RealPath] = append(c.idsByPath[l.RealPath], p.ID)
+		}
+		if l.VirtualPath != "" {
+			c.idsByPath[l.VirtualPath] = append(c.idsByPath[l.VirtualPath], p.ID)
+		}
+	}
+}
+
+func (c *Catalog) Remove(id ID) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	_, exists := c.byID[id]
+	if !exists {
+		log.Errorf("package ID does not exist in the catalog : id=%+v", id)
+		return
 	}
-	c.byType[p.Type] = append(c.byType[p.Type], &p)
 
-	// store by file references
-	for _, s := range p.Locations {
-		_, ok := c.byFile[s]
+	// Remove all index references to this package ID
+	for t, ids := range c.idsByType {
+		c.idsByType[t] = removeID(id, ids)
+		if len(c.idsByType[t]) == 0 {
+			delete(c.idsByType, t)
+		}
+	}
+
+	for p, ids := range c.idsByPath {
+		c.idsByPath[p] = removeID(id, ids)
+		if len(c.idsByPath[p]) == 0 {
+			delete(c.idsByPath, p)
+		}
+	}
+
+	// Remove package
+	delete(c.byID, id)
+}
+
+func (c *Catalog) Finalize() {
+	c.markPackageOwnership()
+}
+
+// markPackageOwnership find overlaps in file ownership with a file that defines another package. Specifically, a .Location.Path of
+// a package is found to be owned by another (from the owner's .Metadata.Files[]). This relationship is captured on the
+// child package.
+func (c *Catalog) markPackageOwnership() {
+	var pkgParents = make(map[ID]*strset.Set)
```

Comment:
I think this function would get a bit simpler and safer if the map's value type wasn't a pointer. This would avoid the need for `nil` checks further on

Reply:
As written in the context of this function the use of a pointer value in a map is safe. (the use of a pointer in of itself is not "unsafe").

Also, the constructor for `strset` provides a pointer to a set, also which nil check are you referring to? The closest I found was:
```
				if _, exists := pkgParents[subPackage.ID]; !exists {
					pkgParents[subPackage.ID] = strset.New()
				}
```
which would be required either way.

Reply:
>the use of a pointer in of itself is not "unsafe"

Agreed.

>which nil check are you referring to?

An imaginary one that's not in the code yet.

As I understand this code, there's no point where a `nil` value can be assigned as a value for a map key in `pkgParents`, so in that sense we're good! The danger is in maintainability and _future_ problems.

Making this a pointer means the system is more complex because there are more possible states in the system, which means it's now less difficult to introduce problems down the road. In this code, there are two accesses of a "potentially" `nil` value (to your point, not today, but there's no guarantee of that remaining true as this code is maintained):

- The call to `.Add` on line 154
- The call `.List` on line 164

If I were a developer in the future coming to work on this function, I would try to look at this code and quickly get a sense for what's going on. I would see that the map's value type was a **pointer** to a `strset.Set`. I'd assume that this was done for a reason, and I'd further assume that the intention was so that the value could possibly be `nil`, since that's the primary impact of writing this as a pointer in the first place. I wouldn't assume that a pointer was added arbitrarily. Given this, I might alter the code in such a way that a `nil` value does become possible. In those cases, the calls to the methods `Add` and `List` would result in nil pointer dereferences. Hopefully, there are tests to catch this. But it would have been a problem that could be avoided altogether.

Reply:
👍 We can move forward with this PR. We'll chat offline and follow up in another PR as necessary.